client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -755,18 +749,65 @@ RequestContext createContextForNextAttempt(Node node, AuthCache authCache) {
         return ignoreErrorCodes;
     }
 
-    private static class ResponseOrException {
+    private static class ResponseOrResponseException {
         private final Response response;
         private final ResponseException responseException;
 
-        ResponseOrException(Response response) {
+        ResponseOrResponseException(Response response) {
             this.response = Objects.requireNonNull(response);
             this.responseException = null;
         }
 
-        ResponseOrException(ResponseException responseException) {
+        ResponseOrResponseException(ResponseException responseException) {
             this.responseException = Objects.requireNonNull(responseException);
             this.response = null;
         }
     }
+
+    /**
+     * Wrap whatever exception we received, copying the type where possible so the synchronous API looks as much as possible
+     * like the asynchronous API. We wrap the exception so that the caller's signature shows up in any exception we throw.
```

Comment:
I think I'd reverse the sentences. "Wrap the exception so the caller's signature shows up in the stack trace, taking care to copy the original type and message where possible so async and sync code don't have to check different exceptions."

Reply:
Or something like that.

Reply:
I'm aware you may be copying a comment that I wrote and I'm effectively commenting on my own way of writing javadoc....

Reply:
:)